rmf_site_format/src/robot_properties.rs

```diff
@@ -399,3 +437,153 @@ impl Recall for RecallCircleCollision {
     }
 }
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "bevy", derive(Component))]
+pub struct PowerSource {
+    pub kind: String,
+    pub config: Value,
+}
+
+impl Default for PowerSource {
+    fn default() -> Self {
+        Self {
+            kind: String::new(),
+            config: Value::Object(Map::new()),
+        }
+    }
+}
+
+#[cfg(feature = "bevy")]
+impl RobotProperty for PowerSource {
+    fn new(kind: String, config: Value) -> Self {
+        Self { kind, config }
+    }
+
+    fn is_default(&self) -> bool {
+        if *self == Self::default() {
+            return true;
+        }
+        false
+    }
+
+    fn kind(&self) -> Option<String> {
+        Some(self.kind.clone())
+    }
+
+    fn label() -> String {
+        "Power Source".to_string()
+    }
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+#[cfg_attr(feature = "bevy", derive(Component))]
+pub struct RecallPowerSource {
+    pub kind: Option<String>,
+    pub config: Option<Value>,
+}
+
+impl Recall for RecallPowerSource {
+    type Source = PowerSource;
+
+    fn remember(&mut self, source: &PowerSource) {
+        self.kind = Some(source.kind.clone());
+        self.config = Some(source.config.clone());
+    }
+}
+
+// Supported kinds of Power Source
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "bevy", derive(Component, Reflect))]
+#[cfg_attr(feature = "bevy", reflect(Component))]
+pub struct Battery {
+    pub voltage: f32,
+    pub capacity: f32,
+    pub charging_current: f32,
+    pub power: f32,
+}
+
+impl Default for Battery {
+    fn default() -> Self {
+        Self {
+            voltage: 12.0,
+            capacity: 24.0,
+            charging_current: 5.0,
+            power: 20.0,
+        }
+    }
+}
+
+#[cfg(feature = "bevy")]
+impl RobotPropertyKind for Battery {
+    fn label() -> String {
+        "Battery".to_string()
+    }
+}
+
+#[cfg(feature = "bevy")]
+impl From<&ElementMap> for Battery {
+    fn from(elements: &ElementMap) -> Self {
+        let mut battery = Battery::default();
+        if let Some(voltage) = elements
+            .get("nominal_voltage")
+            .and_then(|voltage| f64::try_from(voltage.data.clone()).ok())
+        {
+            battery.voltage = voltage as f32;
+        }
+        if let Some(capacity) = elements
+            .get("nominal_capacity")
+            .and_then(|capacity| f64::try_from(capacity.data.clone()).ok())
+        {
+            battery.capacity = capacity as f32;
+        }
+        if let Some(current) = elements
+            .get("charging_current")
+            .and_then(|current| f64::try_from(current.data.clone()).ok())
+        {
+            battery.charging_current = current as f32;
+        }
+        if let Some(power) = elements
+            .get("nominal_power")
+            .and_then(|power| f64::try_from(power.data.clone()).ok())
+        {
+            battery.power = power as f32;
+        }
```

Comment:
Probably similar to my other comment on power, but what does this `nominal_power` actually mean? I'm starting to wonder if we should split the battery parameters into pure battery (voltage, capacity, _most likely_ charging current) and leave all the power parameters such as idle power, tool power, friction coefficient to another component

Reply:
[7fc974f](https://github.com/open-rmf/rmf_site/pull/300/commits/7fc974f79a58c9b45b77fb8fc02d84a4674ab80b). Thanks for pointing this out, I think it's a good call to separate them. I added `PowerDissipation`, `MechanicalSystem` and `AmbientSystem` components to support slotcar parameters, and moved `nominal_power` to `AmbientSystem` as "idle power".

I didn't add in `ToolSystem` this time round since (a) they are not represented in Slotcar params and (b) we might want to think more about what fields to add in if and when we do want to include tool power drain as a robot property kind.

Reply:
I have nothing against using physical battery properties to configure things, but one observation I've made over the years is that system integrators often don't know or care about having accurate parameters for battery properties. They often just think in terms of battery drain rate while moving vs while idle.

We can stick with the physical parameters for now, but down the road we might want to consider just operating in terms of simple "average rate of drain while moving", "average rate of drain while idle", and "average rate of charging" parameters.

Reply:
Would the rate in this case be in Wh or even one higher level and just be a battery percentage per unit time?
But yea I agree that the parameters are a bit hard to get right and something simpler would likely be preferred, we can figure out in a followup how to support both and migrate them